src/platform/backends/qemu/qemu_virtual_machine.cpp

```diff
@@ -212,6 +213,10 @@ std::string mp::QemuVirtualMachine::ipv4()
         while (std::chrono::steady_clock::now() < deadline)
         {
             QCoreApplication::processEvents();
+
+            if (vm_process->state() == QProcess::NotRunning)
```

Comment:
Can we please have a FAILED machine state? Pretty please? :D

And then we can listen on the `QProcess` status events :)

Reply:
Yeah, a FAILED state will make sense for ``start``, but for ``launch``, I'm not so sure.  If we get a failure during ``launch``, we should report the issue back to the client and delete any trace of the instance from the daemon's perspective.

Also, in this particular case, ``qemu-system-x86_64`` just exits, so there in no QProcess error per se.  And we already have a connection to the ``finished`` signal, so we would have to determine in that handler if the process exited because ``stop`` was issued or for some other reasons.

I agree that looking for ``NotRunning`` while polling for various stages during ``launch``/``start`` is not the best, but I think we'll have to re-architecture some things to achieve an event based way of catching things like this.